src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -91,33 +91,50 @@ size_t MonitorList::unlink_deflated(Thread* current, LogStream* ls,
                                     GrowableArray<ObjectMonitor*>* unlinked_list) {
   size_t unlinked_count = 0;
   ObjectMonitor* prev = nullptr;
-  ObjectMonitor* head = Atomic::load_acquire(&_head);
-  ObjectMonitor* m = head;
+  ObjectMonitor* m = Atomic::load_acquire(&_head);
+
   // The in-use list head can be null during the final audit.
   while (m != nullptr) {
     if (m->is_being_async_deflated()) {
-      // Find next live ObjectMonitor.
+      // Find next live ObjectMonitor. Batch up the unlinkable monitors, so we can
+      // modify the list once per batch. The batch starts at "m".
+      size_t unlink_batch = 0;
       ObjectMonitor* next = m;
       do {
         ObjectMonitor* next_next = next->next_om();
         unlinked_count++;
         unlinked_list->append(next);
         next = next_next;
-        if (unlinked_count >= (size_t)MonitorDeflationMax) {
-          // Reached the max so bail out on the gathering loop.
+        if (unlink_batch++ >= (size_t)MonitorUnlinkBatch) {
```

Comment:
Previously we `break` on the `MonitorDeflationMax` which is a implicit limit of the number of deflations. Not as `MonitorUnlinkBatch` could be `MonitorDeflationMax-1` then we would could walk more links than needed. 

Overall I am unsure why we do not pass in deflated_count to unlink and break when the `unlink_count == deflated_count`. Maybe I am missing something and there can be more deflated monitors than  the deflation_count we get. If so it seems like both this and the previous implementation would miss unlinking deflated monitors.

Reply:
Walking more links is fine. Good suggestion to just pass the `deflated_count` here! Let me do that.

Reply:
Actually, now that we have precise `deflated_count`, we might cap the next batch at `deflated_count - unlinked_count` to avoid extra walk, without extra overhead.